### PR TITLE
Completion is not provided inside curly braces settings if the preceeding text contains 'is'

### PR DESCRIPTION
**Pull request by ivanbasov**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/815109

### CODE REVIEW

**Comment by ivanbasov** [comment]
@gafter please let me know if the approach is correct and if we should not fix this on the parser side.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 1]
```diff
@@ -4604,6 +4604,26 @@ class C
...
+    public static void M() 
+    {
+        var a = 1;
+        if (a is { $$})
+    }
+}
+                              </Document>)
+
+                state.SendTypeChars("a")
+                Await state.AssertSelectedCompletionItem(displayText:="a")
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: wait... why?  Why would 'a' be appropriate here?  Isn't this a recursive property pattern?  in which case, i would only expect the members of 'int' to be shown here.  

I don't believe this is a general expression context.

**Comment by CyrusNajmabadi** [comment]
> Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/815109

Can you include the info from that bug?  i cannot get access to it.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 1 above)
Comment (ivanbasov) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: * Yes, currently, only members of `int `are displayed. 
* In 15.9, it was an expression context and displayed `a` and tons of other symbols.
* I understand that the sample in the unit test maybe not an example we need to cover. However, it changed the same way as the one we need:
1. Set a break point
2. Right click on it
3. Choose `Actions`
4. In 'Log a message to Output Window', type "`The value of a is { $$`"
5. Replace $$ with the caret and type `a`

**Expected**
Completion contains `a`

**Actual**
No completion provided

Do you mean we should reduce the scope of the issue to the breakpoint scenario?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: > In 15.9, it was an expression context and displayed a and tons of other symbols.

That seems... bad

> Yes, currently, only members of int are displayed.

That seems.... good :)

> I understand that the sample in the unit test maybe not an example we need to cover. However, it changed the same way as the one we need:

But... why do we want this behavior?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: > In 'Log a message to Output Window', type `The value of a is { $$`

I don't understand what's happening there.  That looks like free-form text.  It's not C#.  What would happen if they wrote: `the value of a has become { $$`?  there's no `is` there, so you won't have a property pattern.

Comment (ivanbasov) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: > But... why do we want this behavior?

We need this behavior in Edit BreakPoint -> Actions -> Log a message to Output Window. I thought that if it changed in all scopes, we may need to fix it in all scopes.

>> In 15.9, it was an expression context and displayed a and tons of other symbols.
> That seems... bad
>
> > Yes, currently, only members of int are displayed.
> That seems.... good :)

I am also not sure that `is {` guarantees the it is followed by a valid property pattern.
Currently, in 16.0, there are two items in the completion popup: `MinValue `and `MaxValue`. Using them leads to the following syntax:
`if (a is { MinValue: 5, MaxValue: 13 })`
which is invalid, as far as I see even in C# 8.0. 

So, why should we restrict the completion to `MinValue `and `MaxValue` only in the case?

**Comment by CyrusNajmabadi** [comment]
> We need this behavior in Edit BreakPoint -> Actions -> Log a message to Output Window. I thought that if it changed in all scopes, we may need to fix it in all scopes.

I don't understand how they're related.  After a `{` is not an expression context.

**Comment by CyrusNajmabadi** [comment]
> I am also not sure that is { guarantees the it is followed by a valid property pattern.
Currently, in 16.0, there are two items in the completion popup: MinValue and MaxValue. Using them leads to the following syntax:
if (a is { MinValue: 5, MaxValue: 13 })
which is invalid, as far as I see even in C# 8.0.

Indeed.  Sounds like it's not filtering to instance properties.  That said, even if we fix this, the position after a `{` in a property-pattern is *not* an expression context.  

I'm happy to discuss this more on gitter if you'd lke.

**Comment by CyrusNajmabadi** [comment]
> I am also not sure that `is {` guarantees the it is followed by a valid property pattern.

I don't understand what `is {` has to do with this issue.  can you give hte information from teh devdiv bug?   All i've seen was a statement of: `The value of a is { $$`

but that's incredibly specious.  That's literally depending on a *human* writing an english phrase that happens to end with `is {` and then hoping it parses a certain way because `is {` is a property pattern in C#.  It makes no sense.  One is english, one is C#.

**Comment by jasonmalinowski** [comment]
Oh maybe there is a difference between the PRs, but in any case we're probably best fixing in one PR and following the test pattern @jcouv is using.

**Comment by gafter** [comment]
From the original issue https://devdiv.visualstudio.com/DevDiv/_workitems/edit/815109 which uses the feature described in https://www.visualmicro.com/page/User-Guide.aspx?doc=Working-With-Breakpoints-When-Hit.html I would expect the contents of the "Log a message to Output Window" form to be treated as the contents of a C# (or VB) interpolated string.

If that is the case, then the curly brace should be properly interpreted by the compiler as the start of an interpolation.

If that is not the case, then I think the implementation of that feature should be reconsidered.

I don't see any correspondence between the tests in this PR and my intuition that we should be looking at an interpolated string.

**Comment by CyrusNajmabadi** [comment]
>I would expect the contents of the "Log a message to Output Window" form to be treated as the contents of a C# (or VB) interpolated string.

Yeah, if that's the case, then we should be explicitly parsing it as such.  We shoudlnt' be taking the raw english text and trying to handle it as if it were C# code.  This is trivial to see if you just replace `The value of a is { $$` with `The value of a became { $$`.  Clearly, the solution here won't work for this, and clearly we're not going to make completion support the user only literally writing something that ends with `is {`.

**Comment by ivanbasov** [comment]
> Oh maybe there is a difference between the PRs, but in any case we're probably best fixing in one PR and following the test pattern @jcouv is using.

Thank you, @jasonmalinowski !

I verified the fix by @jcouv. It actually fixes the issue for `when`
image

I will correct my PR making it similar to #33966 and retarget to master.

**Comment by CyrusNajmabadi** [comment]
> I verified the fix by @jcouv. It actually fixes the issue for when

I odn't understand.  These are unrelated things.  @jcouv was fixing completion in a pattern context.  Your code is not a pattern context.  It's a piece of english text.

**Comment by ivanbasov** [comment]
@CyrusNajmabadi and @gafter, do you mean that the context for the "Log a message to Output Window" is totally incorrect? So, it seems it is not the interpolated string context.

**Comment by CyrusNajmabadi** [comment]
>@CyrusNajmabadi and @gafter, do you mean that the context for the "Log a message to Output Window" is totally incorrect? So, it seems it is not the interpolated string context.

yes. that seems to be a major problem.  Why are we taking *english* text and trying to parse it as C#.  That will never work properly.  Imagine if the user wrote:  `class { $$`.  They shoudl get expression intellisense there, but they likely won't since you'll parse that as a new class declaration".

If this is free-form english text, with holes (i.e., what you would dump to Console.WriteLine), then this should be parsed as a string interpolatoin.

**Comment by ivanbasov** [comment]
In completion we have the following source snapshots
```
    [0..154)     CSharp   V0     [0..154) using System;\r\nusing System.Diagnostics;\r\n\r\n\r\nnamespace ConsoleAppCS\r\n{\r\n    class Program\r\n    {\r\n\r\n        static void Main(string[] args)\r\n        {\r\n;
  [154..174)     CSharp  V36      [0..20) The value of a is {a
  [174..175)      inert   V1       [0..1) ;
  [175..218)     CSharp   V0   [153..196)             var a = 1;\r\n        }\r\n    }\r\n}
```
It seems that the second source snapshot should be 
`  [154..174)     CSharp  V36      [0..20) $”The value of a is {a”`

We should wrap it with `$""` only if we are in `Log a message to Output Window`. Working with the Debugger team, they will inform us about this by providing an extra Role.

**Comment by CyrusNajmabadi** [comment]
@ivanbasov awesome.  Thanks for drilling into this!

**Comment by CyrusNajmabadi** [comment]
> We should wrap it with $"" only if we are in Log a message to Output Window

Note, we should consider if the right thing to do is wrap with `@$""` or `$""`.  For example, do we expect people to be able to write escape sequences in that box and have them respected.  Or if they would be expected to just insert things like newlines.

**Comment by CyrusNajmabadi** [comment]
Moving this PR to draft.  @genlu can you update on if we need to take any steps here, or close out if not.

**Comment by CyrusNajmabadi** [comment]
Closing due to lack of movement.

**Pull request closed without merging**